File: lib/validacion.js
/**
 * DowntimeOS · Reglas de validación de la capa de negocio (Node / ESM).
 * Puerto 1:1 de `server/validacion.py` (PRD, RF-03).
 *
 *   · Campos obligatorios según el origen del lead.
 *   · Correo con formato válido y regla B2B: se rechazan dominios genéricos
 *     (@gmail.com, @hotmail.com, @outlook.com, @yahoo…) cuando la regla está
 *     activa.
 *   · Teléfono/WhatsApp de 10 dígitos (MX) tolerando espacios, guiones y +52.
 *   · Saneamiento: se recortan cadenas y se limita su longitud para que nada
 *     raro llegue a Postgres.
 *
 * La lista de dominios se queda en código (y no en una tabla de Supabase) a
 * propósito: cambia seguido, no vale un round-trip a la base por request, y así
 * el comportamiento es idéntico al del prototipo local.
 */

export const RE_EMAIL = /^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$/;

/** Regla B2B (RF-03): dominios públicos rechazados. */
export const DOMINIOS_GENERICOS = new Set([
  'gmail.com', 'googlemail.com', 'hotmail.com', 'hotmail.es', 'hotmail.mx',
  'outlook.com', 'outlook.es', 'live.com', 'live.com.mx', 'msn.com',
  'yahoo.com', 'yahoo.com.mx', 'yahoo.es', 'icloud.com', 'me.com',
  'aol.com', 'protonmail.com', 'proton.me', 'gmx.com', 'mail.com',
  'zoho.com', 'yandex.com', 'tutanota.com', 'example.com',
]);

// Se puede apagar por entorno sin tocar código (útil para demos).
export const REGLA_B2B_ACTIVA = process.env.REGLA_B2B_ACTIVA !== 'false';

export const LARGO_MAXIMO = 160;
export const ORIGENES_VALIDOS = new Set(['CALCULADORA', 'AUDITORIA']);
export const ESTATUS_VALIDOS = new Set(['NUEVO', 'AUDITORIA_SOLICITADA']);

/** Error de validación con el mapa campo → mensaje que consume el frontend. */
export class ErrorValidacion extends Error {
  constructor(errores) {
    super('Datos inválidos');
    this.name = 'ErrorValidacion';
    this.status = 400;
    this.errores = errores;
  }
}

export function limpiarTexto(valor, maximo = LARGO_MAXIMO) {
  if (valor === null || valor === undefined) return '';
  return String(valor)
    .replace(/\0/g, '')
    .trim()
    .replace(/\s+/g, ' ')
    .slice(0, maximo);
}

/** Tolera espacios, guiones, paréntesis y la lada de país (+52 / +52 1). */
export function normalizarTelefono(valor) {
  let d = String(valor ?? '').replace(/\D/g, '');
  if (d.length === 12 && d.startsWith('52')) d = d.slice(2);
  if (d.length === 13 && d.startsWith('521')) d = d.slice(3);
  return d;
}

export function dominioDe(email) {
  return email.includes('@') ? email.split('@').pop().toLowerCase() : '';
}

export function validarEmail(email, errores, campo = 'email') {
  const limpio = limpiarTexto(email).toLowerCase();

  if (!limpio) {
    errores[campo] = 'El correo corporativo es obligatorio.';
    return limpio;
  }
  if (!RE_EMAIL.test(limpio)) {
    errores[campo] = 'El formato del correo no es válido.';
    return limpio;
  }
  if (REGLA_B2B_ACTIVA && DOMINIOS_GENERICOS.has(dominioDe(limpio))) {
    errores[campo] = 'Usa un correo corporativo. Los dominios públicos '
      + '(@gmail.com, @hotmail.com, @outlook.com...) no son aceptados.';
  }
  return limpio;
}

/** Normaliza el origen recibido; cualquier valor raro cae a CALCULADORA. */
export function normalizarOrigen(valor) {
  const o = String(valor ?? 'CALCULADORA').toUpperCase();
  return ORIGENES_VALIDOS.has(o) ? o : 'CALCULADORA';
}

export function estatusDeOrigen(origen) {
  return origen === 'AUDITORIA' ? 'AUDITORIA_SOLICITADA' : 'NUEVO';
}

/**
 * Valida y normaliza el payload del formulario.
 *
 * @param {object} payload  Cuerpo JSON de la petición.
 * @param {'CALCULADORA'|'AUDITORIA'} origen
 * @returns {object} datos saneados listos para `calcular()` y para insertar.
 * @throws {ErrorValidacion} con `.errores` = { campo: mensaje }
 */
export function validarLead(payload, origen) {
  if (payload === null || typeof payload !== 'object' || Array.isArray(payload)) {
    throw new ErrorValidacion({ _: 'El cuerpo debe ser un objeto JSON.' });
  }

  const errores = {};
  const limpio = {};

  limpio.nombre = limpiarTexto(payload.nombre);
  if (limpio.nombre.length < 3) {
    errores.nombre = 'Escribe tu nombre completo (mínimo 3 caracteres).';
  }

  limpio.empresa = limpiarTexto(payload.empresa);
  if (limpio.empresa.length < 2) {
    errores.empresa = 'El nombre de la empresa es obligatorio.';
  }

  limpio.email = validarEmail(payload.email, errores);

  const telefono = normalizarTelefono(payload.telefono);
  if (!telefono) {
    errores.telefono = 'El teléfono / WhatsApp es obligatorio.';
  } else if (telefono.length !== 10) {
    errores.telefono = 'El teléfono debe tener 10 dígitos.';
  }
  limpio.telefono = telefono;

  limpio.puesto = limpiarTexto(payload.puesto) || 'No especificado';
  limpio.ciudad = limpiarTexto(payload.ciudad);
  limpio.parque_industrial = limpiarTexto(payload.parque_industrial);
  limpio.sector = limpiarTexto(payload.sector);
  limpio.notas = limpiarTexto(payload.notas, 500);

  // El formulario de cierre exige ciudad / parque industrial.
  if (origen === 'AUDITORIA' && !limpio.ciudad) {
    errores.ciudad = 'Indica la ciudad o parque industrial de la planta.';
  }

  // --- Parámetros de la calculadora -----------------------------------------
  // Acepta "1,500" y "$1500" porque el input de la landing es de texto libre.
  const numero = (clave, defecto) => {
    const crudo = payload[clave] ?? defecto;
    const n = Number(String(crudo).replace(/,/g, '').replace(/\$/g, '').trim());
    if (!Number.isFinite(n)) {
      errores[clave] = 'Debe ser un valor numérico.';
      return Number(defecto);
    }
    return n;
  };

  limpio.maquinas = numero('maquinas', 5);
  limpio.turnos = numero('turnos', 2);
  limpio.tarifa_hora = numero('tarifa_hora', 1200);
  limpio.minutos_paro_dia = numero('minutos_paro_dia', 25);
  limpio.divisa = String(payload.divisa ?? 'MXN').toUpperCase() === 'USD' ? 'USD' : 'MXN';

  const utm = (payload.utm && typeof payload.utm === 'object' && !Array.isArray(payload.utm))
    ? payload.utm
    : {};
  limpio.utm = {
    utm_source: limpiarTexto(utm.utm_source) || 'directo',
    utm_medium: limpiarTexto(utm.utm_medium) || 'landing',
    utm_campaign: limpiarTexto(utm.utm_campaign) || 'margen-oculto-2026',
  };

  if (Object.keys(errores).length > 0) throw new ErrorValidacion(errores);
  return limpio;
}

File: local/server/validacion.py
# -*- coding: utf-8 -*-
"""
Reglas de validacion de la Capa 2 (PRD, RF-03).

  * Campos obligatorios segun el origen del lead.
  * Correo con formato valido y regla B2B: se rechazan dominios genericos
    (@gmail.com, @hotmail.com, @outlook.com, @yahoo, etc.) cuando la regla
    esta activa.
  * Telefono/WhatsApp de 10 digitos (MX) tolerando espacios, guiones y +52.
  * Saneamiento: se recortan cadenas y se limita su longitud para que nada
    raro llegue al archivo JSON.
"""
import re

RE_EMAIL = re.compile(r"^[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}$")

# Regla B2B (RF-03): dominios publicos rechazados.
DOMINIOS_GENERICOS = {
    "gmail.com", "googlemail.com", "hotmail.com", "hotmail.es", "hotmail.mx",
    "outlook.com", "outlook.es", "live.com", "live.com.mx", "msn.com",
    "yahoo.com", "yahoo.com.mx", "yahoo.es", "icloud.com", "me.com",
    "aol.com", "protonmail.com", "proton.me", "gmx.com", "mail.com",
    "zoho.com", "yandex.com", "tutanota.com", "example.com",
}

REGLA_B2B_ACTIVA = True

LARGO_MAXIMO = 160
ESTATUS_VALIDOS = {"NUEVO", "AUDITORIA_SOLICITADA"}


class ErrorValidacion(Exception):
    """Agrupa los errores por campo para responder un 400 legible."""

    def __init__(self, errores):
        super().__init__("Datos invalidos")
        self.errores = errores


def limpiar_texto(valor, maximo=LARGO_MAXIMO):
    if valor is None:
        return ""
    texto = str(valor).replace("\x00", "").strip()
    texto = re.sub(r"\s+", " ", texto)
    return texto[:maximo]


def normalizar_telefono(valor):
    digitos = re.sub(r"\D", "", str(valor or ""))
    if digitos.startswith("52") and len(digitos) == 12:
        digitos = digitos[2:]
    if digitos.startswith("521") and len(digitos) == 13:
        digitos = digitos[3:]
    return digitos


def dominio_de(email):
    return email.split("@")[-1].lower() if "@" in email else ""


def validar_email(email, errores, campo="email"):
    email = limpiar_texto(email).lower()
    if not email:
        errores[campo] = "El correo corporativo es obligatorio."
        return email
    if not RE_EMAIL.match(email):
        errores[campo] = "El formato del correo no es válido."
        return email
    if REGLA_B2B_ACTIVA and dominio_de(email) in DOMINIOS_GENERICOS:
        errores[campo] = (
            "Usa un correo corporativo. Los dominios públicos "
            "(@gmail.com, @hotmail.com, @outlook.com...) no son aceptados."
        )
    return email


def validar_lead(payload, origen):
    """Valida y normaliza el payload del formulario.

    `origen` es "CALCULADORA" (lead magnet / reporte PDF) o "AUDITORIA"
    (formulario de cierre de 30 dias). Lanza ErrorValidacion si algo falla.
    """
    if not isinstance(payload, dict):
        raise ErrorValidacion({"_": "El cuerpo debe ser un objeto JSON."})

    errores = {}
    limpio = {}

    limpio["nombre"] = limpiar_texto(payload.get("nombre"))
    if len(limpio["nombre"]) < 3:
        errores["nombre"] = "Escribe tu nombre completo (mínimo 3 caracteres)."

    limpio["empresa"] = limpiar_texto(payload.get("empresa"))
    if len(limpio["empresa"]) < 2:
        errores["empresa"] = "El nombre de la empresa es obligatorio."

    limpio["email"] = validar_email(payload.get("email"), errores)

    telefono = normalizar_telefono(payload.get("telefono"))
    if not telefono:
        errores["telefono"] = "El teléfono / WhatsApp es obligatorio."
    elif len(telefono) != 10:
        errores["telefono"] = "El teléfono debe tener 10 dígitos."
    limpio["telefono"] = telefono

    limpio["puesto"] = limpiar_texto(payload.get("puesto")) or "No especificado"
    limpio["ciudad"] = limpiar_texto(payload.get("ciudad"))
    limpio["parque_industrial"] = limpiar_texto(payload.get("parque_industrial"))
    limpio["sector"] = limpiar_texto(payload.get("sector"))
    limpio["notas"] = limpiar_texto(payload.get("notas"), 500)

    # El formulario de cierre exige ciudad / parque industrial.
    if origen == "AUDITORIA" and not limpio["ciudad"]:
        errores["ciudad"] = "Indica la ciudad o parque industrial de la planta."

    # --- Parametros de la calculadora -------------------------------------
    def numero(clave, defecto):
        crudo = payload.get(clave, defecto)
        try:
            return float(str(crudo).replace(",", "").replace("$", "").strip())
        except (TypeError, ValueError):
            errores[clave] = "Debe ser un valor numérico."
            return float(defecto)

    limpio["maquinas"] = numero("maquinas", 5)
    limpio["turnos"] = numero("turnos", 2)
    limpio["tarifa_hora"] = numero("tarifa_hora", 1200)
    limpio["minutos_paro_dia"] = numero("minutos_paro_dia", 25)
    limpio["divisa"] = "USD" if str(payload.get("divisa", "MXN")).upper() == "USD" else "MXN"

    utm = payload.get("utm") if isinstance(payload.get("utm"), dict) else {}
    limpio["utm"] = {
        "utm_source": limpiar_texto(utm.get("utm_source")) or "directo",
        "utm_medium": limpiar_texto(utm.get("utm_medium")) or "landing",
        "utm_campaign": limpiar_texto(utm.get("utm_campaign")) or "margen-oculto-2026",
    }

    if errores:
        raise ErrorValidacion(errores)
    return limpio
